### DISCUSSION TITLE
Adding custom icons to toolbar

### DISCUSSION THREAD

**Discussion by Tenderenden** [sumatrapdfreader/sumatrapdf#3339] [Q&A] [answered]
Hello,
I'm trying to add custom icon to toolbar to enable toggling bookmarks on/off. I chose [this icon](https://github.com/tabler/tabler-icons/blob/master/icons/bookmarks.svg) which seems to be from the same repository as the other icons. I found that icons used for toolbar are stored as svg strings in file `SvgIcons.cpp` so I tried to add one by copy-pasting text content of my svg file and then adding it in `gAllIcons` variable so I can later use it in `gToolbarButtons` in `Toolbar.cpp`. That method didn't work so I tried to copy existing string `gIconRotateRight` and paste it as my custom icon. This way it worked and I got additional icon on toolbar with rotate icon.

My questions are: how those SVG strings should be modified to work properly? Should I add my custom icon in other place as well?

**Comment by kjk** [comment]
It would be easier to answer if you posted a link to a repository with changes you made.

Generally speaking, as you found out, toolbar code is flexible enough to easily change the icon with a new SVG.

But it's not flexible enough to make it easy to add new icons and connect them with functionality.

Currently the number and order of toolbar icons / elements is hardcoded.

You can add new icons to the toolbar but you'll have to make more changes to `Toolbar.cpp`, `gToolbarButtons` etc.

It's possible to make it customizable via settings file but other things were more important so far.

**Comment by Tenderenden** [comment, marked as answer]
Thanks for quick reply. After your advice I created new fork of repo and started recreating my solution step by step and for some reason it worked this time. Even though I got some build and linking errors. I probably didn't clear build directory before rebuilding and these errors made me belive it didn't work.
Now I have new icon to toggle bookmarks :)
![obraz](https://user-images.githubusercontent.com/20031157/224391273-d27d0a11-55f1-42b9-b53a-0afec83dde0a.png)

All the changes are in my fork on branch [Add-toolbar-icon](https://github.com/Tenderenden/sumatrapdf-Tenderenden/tree/Add-toolbar-icon)

Thanks for your help!
